reactor-extra/src/main/java/reactor/retry/DefaultRepeat.java

```diff
@@ -0,0 +1,139 @@
+
+import java.time.Duration;
+import java.util.function.Consumer;
+import java.util.function.Function;
+import java.util.function.Predicate;
+
+import org.reactivestreams.Publisher;
+
+import reactor.core.publisher.Flux;
+import reactor.core.scheduler.Scheduler;
+import reactor.util.Logger;
+import reactor.util.Loggers;
+
+public class DefaultRepeat<T> extends AbstractRetry<T, Long> implements Repeat<T> {
+
+	static final Logger log = Loggers.getLogger(DefaultRepeat.class);
+	static final Consumer<? super RepeatContext<?>> NOOP_ON_REPEAT = r -> {};
+
+	final Predicate<? super RepeatContext<T>> repeatPredicate;
+	final Consumer<? super RepeatContext<T>> onRepeat;
+
+	DefaultRepeat(Predicate<? super RepeatContext<T>> repeatPredicate,
+			int maxRepeats,
+			Duration timeout,
+			Backoff backoff,
+			Jitter jitter,
+			Scheduler backoffScheduler,
+			final Consumer<? super RepeatContext<T>> onRepeat,
+			T applicationContext) {
+		super(maxRepeats, timeout, backoff, jitter, backoffScheduler, applicationContext);
+		this.repeatPredicate = repeatPredicate;
+		this.onRepeat = onRepeat;
+	}
+
+	public static <T> DefaultRepeat<T> create(Predicate<? super RepeatContext<T>> repeatPredicate, int n) {
+		return new DefaultRepeat<T>(repeatPredicate,
+				n,
+				null,
+				Backoff.zero(),
+				Jitter.noJitter(),
+				null,
+				NOOP_ON_REPEAT,
+				(T) null);
+	}
+
+	@Override
+	public Publisher<Long> apply(Flux<Long> companionValues) {
+		return new RepeatFunction().apply(companionValues);
+	}
+
+	@Override
+	public Repeat<T> withApplicationContext(T applicationContext) {
+		return new DefaultRepeat<>(repeatPredicate, maxIterations, timeout,
+				backoff, jitter, backoffScheduler, onRepeat, applicationContext);
+	}
+
+	@Override
+	public Repeat<T> doOnRepeat(Consumer<? super RepeatContext<T>> onRepeat) {
+		return new DefaultRepeat<>(repeatPredicate, maxIterations, timeout,
+				backoff, jitter, backoffScheduler, onRepeat, applicationContext);
+	}
+
+	@Override
+	public Repeat<T> timeout(Duration timeout) {
+		if (timeout.isNegative())
+			throw new IllegalArgumentException("timeout should be >= 0");
+		return new DefaultRepeat<>(repeatPredicate, Integer.MAX_VALUE, timeout,
+				backoff, jitter, backoffScheduler, onRepeat, applicationContext);
+	}
+
+	@Override
+	public Repeat<T> backoff(Backoff backoff) {
+		return new DefaultRepeat<>(repeatPredicate, maxIterations, timeout,
+				backoff, jitter, backoffScheduler, onRepeat, applicationContext);
+	}
+
+	@Override
+	public Repeat<T> jitter(Jitter jitter) {
+		return new DefaultRepeat<>(repeatPredicate, maxIterations, timeout,
+				backoff, jitter, backoffScheduler, onRepeat, applicationContext);
+	}
+
+	@Override
+	public Repeat<T> withBackoffScheduler(Scheduler scheduler) {
+		return new DefaultRepeat<>(repeatPredicate, maxIterations, timeout,
+				backoff, jitter, scheduler, onRepeat, applicationContext);
+	}
+
+	class RepeatFunction implements Function<Flux<Long>, Publisher<Long>> {
```

Comment:
We basically just need somewhere to store the previous backoff. That is the only mutable state. Moved that state to DefaultContext and removed the inner class.